Modernize Makefile

erbsenzaehler opened the pull request:
I did some work on the Makefile and added some targets. Maybe you like it. If this gets commited, you can remove the now unused compile.linux.

tico-tico commented:
Hi, erbsenzaehler, thanks for your help.

1. *line 20*
`-msse` will it restrict to SSE for x64 too (i have no idea)? There are no x64 without SSE2.
And I think you want `-mfpmath=sse` too.

2. *line 20, 21*
Why annoy people with `-Wextra`? I'm just afraid that someone will start to "fix" the shouldn't-be-fixed warnings like this one `‘best_pv’ may be used uninitialized...`. If it's uninitialized then something really really went wrong and i prefer a loud crash in this case.

3. *line 49*
`-msse3 -mpopcnt`
Isn't POPCNT a part of SSE4.2? Why restrict to SSE3?

4. *lines 72, 74*
PGO doesn't require `-lgcov` here in GCC7 on Arch.

5. *lines 135*
`Build a binary with Logfile support`. What is "Logfile support"?

6. Are you sure about `bench 12`? Is it possible that is will slow down functions not involved in the benchmark? Maybe just `bench` would be enough.

@alwey will this *Makefile* work for your ancient GCC? (i hope this `@name` thing actually sends notifications).

tico-tico commented:
Almost forgot, It would be great if you create a `hyper-speed` target with `PGO` + `-march=native` for i-will-build-it-myself people.

alwey commented:
@erbsenzaehler  @tico-tico : I must try this, My first impression: it will break things, for various reasons.

alwey commented:
@erbsenzaehler  @tico-tico : actually it seems to work better now.

erbsenzaehler commented:
@alwey  What do you mean by it  seems to work better now?

@tico-tico 

1.) I did some quick measurements and it there is no difference with -msse or without, so I removed it.

2) I have no strong opinion on this, so I removed it.

3) -mpopcnt -msse3 works on some older amd cpus, which do not have msse4.2, yet

4) -fprofile-generate and -fprofile-use implicitly call -lgcov, so -lgcov can be removed

5) Logfile support was in the Makefile before. But I could not find any reference to it in the code, so it is removed.

6) I did a quick measure with bench 12 and bench and again there was no difference speed-wise. So stick with the simpler one. I added lcov coverage similar to https://github.com/official-stockfish/Stockfish/pull/1181/files

You can now add native=yes to your command line and it will compile with -march=native.

alwey commented:
@erbsenzaehler The new Makefile is functional. It has a better structure and can be adapted more easily.

tico-tico commented:
If everyone is happy then, @nescitus Pawel, I think it's your turn to merge the pull request.
